File: HW2/Q1.cpp
#include<iostream>
#include<cmath>
#include<cstdlib>   // 亂數相關函數
#include<ctime>     // 時間相關函數
#include<vector>
#include<climits>
using namespace std;

struct cell{
    double x1, x2;
    double fitness;
};

class GAFloat{
public:
    int max_iter, population_size;
    float p_mutation, p_crossover;
    vector<cell> population;   // The total number of population.
    vector<cell> pool;
    vector<int> best_gene_list;
    cell best_cell;
    int best_iter, cur_iter=0;
    GAFloat(int max_iter, int population_size, float min_bound, float max_bound, float precision, float p_mutation, float p_crossover);
    void initialize();
    void evaluate();
    void crossover();
    void mutation();
    void select(string mode, int times);
    void run(string mode, int times);
    int find_best(string mode);
    void print_info(int iter_interval);
};

GAFloat::GAFloat(int max_iter, int population_size, float min_bound, float max_bound, float precision, float p_mutation, float p_crossover){

    this->max_iter = max_iter;
    this->population_size = population_size;
    this->p_mutation = p_mutation;
    this->p_crossover = p_crossover;

    this->p_mutation = p_mutation;
    this->p_crossover = p_crossover;

    cout<<"Constructor.\n";
}

void GAFloat::initialize(){
    for(int i=0;i<population_size;i++){
        cell node;
        node.x1 = (double)rand() / RAND_MAX;
        node.x2 = (double)rand() / RAND_MAX;
        population.push_back(node);
    }
}

void GAFloat::evaluate(){
    for(auto& node: population){
        double x1=node.x1, x2=node.x2;
        double left_part = pow(x1*x1 + x2*x2, 0.25),
                right_part = pow(sin(50*pow((x1*x1 + x2*x2), 0.1)), 2.0) + 1;

        node.fitness = left_part * right_part;
    }
}

void GAFloat::select(string mode, int times){
    pool.clear();
    for(int i=0;i<population_size;i++){
        int select_idx = rand() % population_size;
        for(int j=0;j<times;j++){
            int idx = rand() % population_size;
            if(mode == "max"){
                if(population[select_idx].fitness < population[idx].fitness)
                    select_idx = idx;
            }
            if(mode == "min"){
                if(population[select_idx].fitness > population[idx].fitness)
                    select_idx = idx;
            }
        }
        pool.push_back(population[select_idx]);
    }
    population.clear();
    population.assign(pool.begin(), pool.end());
}

void GAFloat::crossover(){
    int idx1, idx2;
    for(int i=0;i<population_size;i++){
        if((double)rand() / RAND_MAX > p_crossover)  // Do not corssover.
            continue;
            idx1 = rand() % population_size;
            idx2 = rand() % population_size;
            while(idx2 == idx1){
                idx2 = rand() % population_size;

            population[idx1].x2 = pool[idx2].x2;
            population[idx2].x1 = pool[idx1].x1;
        }
    }
}

void GAFloat::mutation(){
    for(auto& node: population){
        if((double)rand() / RAND_MAX < p_mutation)
            node.x1 = rand() / RAND_MAX;
        if((double)rand() / RAND_MAX < p_mutation)
            node.x2 = rand() / RAND_MAX;
    }
}

void GAFloat::run(string mode, int times){
    best_cell.fitness = (mode == "max") ? INT_MIN : INT_MAX;
    initialize();
    evaluate();
    for(int iter=0;iter<max_iter;iter++){
        select(mode, times);
        crossover();
        mutation();
        evaluate();
        int best_idx = find_best(mode);
        best_gene_list.push_back(best_idx);
    }
    int iter_interval = 200;
    print_info(iter_interval);
}

int GAFloat::find_best(string mode){
    cur_iter++;
    int idx;
    if(mode == "max"){
        double maximum = INT_MIN;
        for(int i=0;i<population_size;i++){
            if(population[i].fitness > maximum){
                maximum = population[i].fitness;
                idx = i;
                if(best_cell.fitness < maximum){
                    best_cell = population[i];
                    best_iter = cur_iter;
                }
            }
        }
    }
    if(mode == "min"){
        double minimum = INT_MAX;
        for(int i=0;i<population_size;i++){
            if(population[i].fitness < minimum){
                minimum = population[i].fitness;
                idx = i;
                if(best_cell.fitness > minimum){
                    best_cell = population[i];
                    best_iter = cur_iter;
                }
            }
        }
    }
    return idx;
}

void GAFloat::print_info(int iter_interval){
    int iter = 1;
    for(int iter=0;iter<best_gene_list.size();iter+=iter_interval){
        int idx = best_gene_list[iter];
        double x1_val=population[idx].x1,
            x2_val=population[idx].x2;
        cout<<"Iteration: "<<iter<<"\n"
            <<"best fitness: "<<population[idx].fitness<<"\n"
            <<"(x1, x2) = ("<<x1_val<<", "<<x2_val<<")\n";

        cout<<"======\n";
    }
    double x1_val=best_cell.x1, x2_val=best_cell.x2;

    cout<<"All best fitness: "<<best_cell.fitness<<endl;
    cout<<"All best (x1, x2): "<<x1_val<<", "<<x2_val<<endl;
    cout<<"All best iter: "<<best_iter<<endl;
}

int main(){
    srand((unsigned)time(NULL));  // (unsigned)time(NULL)

    int max_iter=10000, population_size=100, min_bound=0, max_bound=1;
    float precision=0.0001, p_mutation=0.01, p_crossover=0.25;
    

    string mode = "max";
    int times = 5;

    // 收集實驗數據用於計算平均和最大最小值範圍
    vector<float> total_fitness, total_iter, total_x1, total_x2;
    // 收集數據之實驗次數
    int exp_num=1;
    for(int exp=0;exp<exp_num;exp++){
        // Construct instance;
        GAFloat ga(
        max_iter,
        population_size,
        min_bound,
        max_bound,
        precision,
        p_mutation,
        p_crossover);

        ga.run(mode, times);

        total_fitness.push_back(ga.best_cell.fitness);
        total_x1.push_back(ga.best_cell.x1);
        total_x2.push_back(ga.best_cell.x2);
        total_iter.push_back(ga.best_iter);
        
    }


    cout<<"===\n";
    cout<<"|name|stats|\n";
    cout<<"|-|-|\n";
    cout<<"|fitness mean| "<<setprecision(6)<<find_sum(total_fitness)/ exp_num<<"|\n";
    cout<<"|iter mean   | "<<find_sum(total_iter) / exp_num<<"|\n";
    cout<<"|x1 mean     | "<<setprecision(4)<<find_sum(total_x1) / exp_num<<"|\n";
    cout<<"|x2 mean     | "<<setprecision(4)<<find_sum(total_x2) / exp_num<<"|\n";

    find_range(total_fitness, "fitness");
    find_range(total_iter, "iter");
    find_range(total_x1, "x1");
    find_range(total_x2, "x2");

    cout<<"\n Now mode: "<<mode<<endl;
    cout<<"GA float string\n";
}

File: HW2/Q1.py
import copy
from math import sqrt, cos, sin, pi
import random
import sys

import numpy as np


def fitness_func(x, y):
    return  80 - x**2 - y**2 + 10*cos(2*pi*x) + 10*cos(2*pi*y)

class Cell:
    def __init__(self, x, y, fitness=0.0):
        self.x = x
        self.y = y
        self.fitness = fitness

class NichGA:
    def __init__(
        self,
        population_size: int,
        max_iter: int,
        p_crossover: float,
        p_mutation: float,
        radius: float,
        alpha: float,
        lower: float,
        upper: float,
        select_times: int
    ):
        r"""[GA with Niching sharing method.]

        Args:
            radius (float): [The threshold of dissimilarity (also the
                            distance cutoff or the niche radius]
            alpha (float): [Regulates the shape of the sharing function.]
        """
        self.population_size = population_size
        self.max_iter = max_iter
        self.p_crossover = p_crossover
        self.p_mutation = p_mutation
        self.radius = radius
        self.alpha = alpha
        self.select_times = select_times

        self.lower = lower
        self.upper = upper

        self.population = []
        self.pool = []
        self.best_gene_list = []
        self.best_cell = Cell(0, 0, -sys.maxsize)

    def initialize(self):
        for _ in range(self.population_size):
            x = random.uniform(self.lower, self.upper)
            y = random.uniform(self.lower, self.upper)
            # fitness = fitness_func(x, y)
            self.population.append(Cell(x, y, 0))

    def compute_niche_count(self, cur_node):
        niche_count = 0.0
        for node in self.population:
            dis = self.distance(cur_node, node)
            if dis < self.radius:
                niche_count += 1 - (dis / self.radius)**self.alpha
        
        return niche_count if niche_count != 0 else 1.0

    def distance(self, src, tar):
        return sqrt((src.x - tar.x)**2 + (src.y - tar.y)**2)

    def evaluate(self, mode='no'):
        for node in self.population:
            node.fitness  = fitness_func(node.x, node.y)
        if mode == 'init':
            return

        for node in self.population:
            niche_count = self.compute_niche_count(node)
            node.fitness /= niche_count

    def select(self):
        self.pool.clear()
        for node in self.population:
            select_idx = random.randint(0, self.population_size-1)
            for _ in range(self.select_times):
                idx = random.randint(0, self.population_size-1)
                if self.population[select_idx].fitness < self.population[idx].fitness:
                    select_idx = idx

            self.pool.append(self.population[select_idx])

        self.population.clear()
        self.population = copy.deepcopy(self.pool)
    
    def crossover(self):
        for _ in range(self.population_size):
            if random.uniform(0, 1) > self.p_crossover:
                continue
            idx1 = random.randint(0, self.population_size-1)
            idx2 = random.randint(0, self.population_size-1)
            while idx1 == idx2:
                idx2 = random.randint(0, self.population_size-1)

            self.population[idx1].x = self.population[idx2].x
            self.population[idx2].y = self.population[idx1].y

    def mutation(self):
        for node in self.population:
            if random.uniform(0, 1) < self.p_mutation:
                node.x = random.uniform(self.lower, self.upper)
            if random.uniform(0, 1) < self.p_mutation:
                node.y = random.uniform(self.lower, self.upper)

    def run(self):
        self.initialize()
        self.evaluate('init')
        for iter in range(self.max_iter):
            self.select()
            self.crossover()
            self.mutation()
            self.evaluate()

            # Record the best gene in each iteration.
            cur_best_gene = self.find_best()
            self.best_gene_list.append(cur_best_gene)

        self.display_info()

    def find_best(self):
        self.real_eval()
        maximum = -sys.maxsize
        cur_best_iter, cur_best_idx = 0, 0
        for iter, idx in enumerate(range(self.population_size)):
            if self.pool[idx].fitness >  maximum:
                maximum = self.pool[idx].fitness
                cur_best_idx = idx
                cur_best_iter = iter

        # Update the best cell & best iter.
        if self.best_cell.fitness < maximum:
            self.best_cell = self.pool[cur_best_idx]
            self.best_iter = cur_best_iter

        return self.pool[cur_best_idx]


    def display_info(self, iter_interval: int = 100):
        for iter in range(0, len(self.best_gene_list), iter_interval):
            node = self.best_gene_list[iter]
            print(f'Iteration: {iter}')
            print(f'Best fitness: {node.fitness}')
            print(f'(x, y): ({node.x}, {node.y})')
            print('========\n')

        print(f'All best fitness: {self.best_cell.fitness}')
        print(f'All best (x, y): ({self.best_cell.x}, {self.best_cell.y})')
        print(f'All best iter: {self.best_iter}')

    def real_eval(self):
        for node in self.pool:
            node.fitness = fitness_func(node.x, node.y)

    def find_peak(self, k: int = 3):
        top_k = 4
        tmp = []
        print(f'\nFind {k} peaks')
        sorted_genes = sorted(self.best_gene_list, key=lambda x: x.fitness, reverse=True)
        for i, node in enumerate(sorted_genes[:k]):
            # print(f'Peak {i}:')
            # print(f'fitness {node.fitness}:')
            # print(f'(x, y) ({node.x:.4f}, {node.y:.4f}):')
            # print('----')
            # if i < top_k or node.fitness >= 91:
            #     tmp.append(node)
            #     continue

            counter = 0
            for other in tmp:
                if self.distance(other, node) >= 1:
                    print('##')
                    counter += 1

            if counter == len(tmp):
                tmp.append(node)
            
            if len(tmp) == top_k
        for i, node in enumerate(tmp):
            print(f'Peak {i}')
            print(f'fitness = {fitness_func(node.x, node.y)}:')
            print(f'(x, y) = ({node.x:.4f}, {node.y:.4f}):')
            print('----')

        return tmp


if __name__ == '__main__':
    max_iter = 1000
    population_size = 150
    lower, upper = -0.5, 1.500001
    p_mutation, p_crossover = 0.01, 0.25
    select_times = 5

    alpha = 1
    radius = 0.25

    exp_num = 1
    for exp in range(exp_num):
        ga = NichGA(
            population_size=population_size,
            max_iter=max_iter,
            p_crossover=p_crossover,
            p_mutation=p_mutation,
            radius=radius,
            alpha=alpha,
            lower=lower,
            upper=upper,
            select_times=select_times
        )

        ga.run()

        ga.find_peak(k=30)
    